File: backend/app/models/modem.py
from sqlalchemy import Column, Integer, String, Boolean, DateTime, Text, Enum
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from app.core.database import Base


class ModemStatus(str, enum.Enum):
    CONNECTED = "connected"
    DISCONNECTED = "disconnected"
    ERROR = "error"
    UNKNOWN = "unknown"


class Modem(Base):
    __tablename__ = "modems"

    id = Column(Integer, primary_key=True, index=True)
    device_path = Column(String(100))
    # mm_object_path 是设备唯一键，格式：
    #   标准 mmcli 设备：/org/freedesktop/ModemManager1/Modem/<n>
    #   ZTE 随身 WiFi：  zte:192.168.0.1（合成路径，非 D-Bus）
    # 不要用 device_path 或 imei 作唯一标识（可能为空或在设备重启后变更）
    mm_object_path = Column(String(200), unique=True, nullable=False)
    imei = Column(String(20), unique=True)
    manufacturer = Column(String(100))
    model = Column(String(100))
    phone_number = Column(String(30))
    operator = Column(String(100))
    signal_quality = Column(Integer, default=0)
    status = Column(Enum(ModemStatus), default=ModemStatus.UNKNOWN)
    alias = Column(String(100))
    is_active = Column(Boolean, default=True)
    last_seen = Column(DateTime, default=datetime.utcnow)
    created_at = Column(DateTime, default=datetime.utcnow)

    # Extended stats (populated by poller)
    access_technologies = Column(String(100))   # e.g. "lte", "umts"
    registration_state = Column(String(50))     # e.g. "home", "roaming"
    tx_bytes = Column(Integer, default=0)       # bytes sent via bearer
    rx_bytes = Column(Integer, default=0)       # bytes received via bearer
    connection_duration = Column(Integer, default=0)  # seconds

    # SIM card details
    imsi = Column(String(20))
    iccid = Column(String(30))
    firmware_revision = Column(String(100))
    hardware_revision = Column(String(50))
    current_bands = Column(String(500))         # comma-separated band list
    sim_operator_name = Column(String(100))     # SIM card operator name (e.g. giffgaff)
    sim_operator_code = Column(String(20))      # MCC+MNC (e.g. 23410)
    current_modes = Column(String(200))         # e.g. "allowed: 2g, 3g, 4g; preferred: 4g"
    ports = Column(String(300))                 # comma-separated port list
    plugin = Column(String(50))                 # MM plugin name (e.g. quectel)

    sms_messages = relationship("SmsMessage", back_populates="modem")
    scheduled_tasks = relationship("SmsScheduledTask", back_populates="modem")

File: backend-go/models/modem.go
package models

import "time"

// Modem 状态常量，由轮询器写入并通过 WebSocket 推送给前端。
const (
	ModemConnected    = "connected"    // 已注册/连接
	ModemDisconnected = "disconnected" // 断线/禁用/搜网中
	ModemError        = "error"        // 故障状态
	ModemUnknown      = "unknown"      // 状态未知
)

// Modem 表示一个 USB 调制解调器（或 ZTE 便携 WiFi 设备）的持久化状态。
// MmObjectPath 是唯一键：mmcli D-Bus 路径，如 /org/freedesktop/ModemManager1/Modem/0；
// ZTE 设备使用合成路径 zte:192.168.0.1。
type Modem struct {
	ID           uint   `gorm:"primaryKey" json:"id"`
	DevicePath   string `gorm:"size:100" json:"device_path"`           // 设备节点，如 ttyUSB0
	MmObjectPath string `gorm:"column:mm_object_path;size:200;uniqueIndex;not null" json:"mm_object_path"` // 唯一键
	Imei         string `gorm:"size:20;uniqueIndex" json:"imei"`       // IMEI，可能为空
	Manufacturer string `gorm:"size:100" json:"manufacturer"`
	Model        string `gorm:"size:100" json:"model"`
	PhoneNumber  string `gorm:"size:30" json:"phone_number"`
	Operator     string `gorm:"size:100" json:"operator"` // 运营商名称

	SignalQuality int       `gorm:"default:0" json:"signal_quality"` // 信号质量 0-100
	Status        string    `gorm:"size:20;default:unknown" json:"status"`
	Alias         string    `gorm:"size:100" json:"alias"` // 用户自定义别名
	IsActive      bool      `gorm:"default:true" json:"is_active"` // false 表示设备已拔出
	LastSeen      time.Time `json:"last_seen"`
	CreatedAt     time.Time `json:"created_at"`

	AccessTechnologies string `gorm:"size:100" json:"access_technologies"` // 接入技术，如 lte
	RegistrationState  string `gorm:"size:50" json:"registration_state"`
	TxBytes            int64  `gorm:"default:0" json:"tx_bytes"`
	RxBytes            int64  `gorm:"default:0" json:"rx_bytes"`
	ConnectionDuration int64  `gorm:"default:0" json:"connection_duration"` // 秒

	Imsi             string `gorm:"size:20" json:"imsi"`
	Iccid            string `gorm:"size:30" json:"iccid"`
	FirmwareRevision string `gorm:"size:100" json:"firmware_revision"`
	HardwareRevision string `gorm:"size:50" json:"hardware_revision"`
	CurrentBands     string `gorm:"size:500" json:"current_bands"`
	SimOperatorName  string `gorm:"size:100" json:"sim_operator_name"` // SIM 卡归属运营商名
	SimOperatorCode  string `gorm:"size:20" json:"sim_operator_code"`  // MCC+MNC
	CurrentModes     string `gorm:"size:200" json:"current_modes"`
	Ports            string `gorm:"size:300" json:"ports"`
	Plugin           string `gorm:"size:50" json:"plugin"` // mmcli 插件名
}

func (Modem) TableName() string { return "modems" }
